analyzers/src/SonarAnalyzer.Common/Rules/UseLambdaParameterInConcurrentDictionaryBase.cs

```diff
@@ -0,0 +1,66 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class UseLambdaParameterInConcurrentDictionaryBase<TSyntaxKind, TInvocationExpression, TArgumentSyntax> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+    where TInvocationExpression : SyntaxNode
+    where TArgumentSyntax : SyntaxNode
+{
+    private const string DiagnosticId = "S6612";
+
+    protected override string MessageFormat => "Use the lambda parameter instead of capturing the argument";
+
+    protected abstract SeparatedSyntaxList<TArgumentSyntax> GetArguments(TInvocationExpression invocation);
+    protected abstract bool TryGetKeyName(TArgumentSyntax argument, out string keyName);
+    protected abstract bool IsLambdaAndContainsIdentifier(string keyName, TArgumentSyntax argument);
+
+    protected UseLambdaParameterInConcurrentDictionaryBase() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(Language.GeneratedCodeRecognizer, c =>
+        {
+            var invocation = (TInvocationExpression)c.Node;
+
+            if (HasCorrectName(Language.GetName(invocation))
+                && Language.Syntax.TryGetOperands(invocation, out var left, out var right)
+                && IsCorrectType(left, c.SemanticModel)
+                && c.SemanticModel.GetSymbolInfo(right).Symbol.ContainingNamespace.Name == "Concurrent"
```

Comment:
Educational: I understand this check ensures the method is part of `ConcurrentDictionnary`.
Why not check if the method is part of the `ConcurrentDictionnary`?
Like we do to check for extensions method but for actual methods:
```cs
protected static bool IsCorrectCall(SyntaxNode right, SemanticModel model) =>
    model.GetSymbolInfo(right).Symbol is IMethodSymbol method
    && method.IsExtensionOn(KnownType.System_Collections_Generic_IEnumerable_T);
```

Reply:
If you are worried about people "hiding" the method, you can check if it's defined inside `ConcurrentDictionary` explicitly by getting the `IMethodSymbol` of `right` and then check:
```csharp
methodSymbol.ContainingType.Is(KnownType.ConcurrentThingy);
```

This should not raise in this case:
```csharp
class HidesMethod<TKey, TValue> : ConcurrentDictionary<TKey, TValue>
{
    // hides the method you are looking for
    public TValue GetOrAdd(TKey key, Func<TKey, TValue> valueFactory) => default(TValue);
}
```

And add a couple of UTs that cover this weird case.